### PR TITLE
feature #87: create admin students flow

### LINKED ISSUE #87: [FEATURE] Create an admin students flow
We need to create "Search students", "Edit student", "Create student" pages

## Front-end part

Search students:

- Sidebar that is already implemented
- Search panel:
  - Search input that searches by full name of student
  - Filter by group, multichoice dropdown with an search input
  - Filter by group role, multichoice dropdown with an search input
  - Sorting input with variants "by name", "by group", "by group role"
  - Icon button to change the direction of sorting (asc/desc)
- Button "Create" which opens "Create student" page
- Table:
  - student name (string)
  - group code
  - group role (in a form of a tag, CAPTAIN, STUDENT, MODERATOR only)
  - Button "Edit" which opens "Edit department" page
  - Button "Delete" which opens confirmation pop-up
  - Below the table there are settings for page size and page (from MUI component)

image

Edit student:

- Sidebar that is already implemented
- Button "Cancel" which opens "Search students" page
- Button "Delete" which opens confirmation pop-up. After deleting opens "Search students" page
- Button "Save" which send a request on back-end endpoint which saves data. After saving opens "Search students" page
- Input for first name (min 2, max 40)
- Input for last name (min 2, max 40)
- Input for middle name (min 0, max 40)
- Dropdown for group, multichoice dropdown with an search input
- Dropdown for group role (CAPTAIN, STUDENT, MODERATOR only)
- Button "Edit user" that opens "Edit user" page with user id that has student
- Edit selectives:
  - Each part splitted by semester where user can have selectives:
    - In each part you can add or delete selectives:
      - Bucket icon will delete selectives
      - Button "Add" will show new dropdown, where you can select new discipline. In these dropdown there cannot be disciplines that already picked by the user
      - Also we will send infromation about max amount of selective that user can have in each semester. If limit is reached, do not show "Add" button for exact semester
  - If group has been changed, selective disciplines must be empty
  - If group has been changed, use endpoint for getting selective disciplines from group (see endpoints below)
  - If group has not been changed, use endpoint for getting remaining selective disciplines (see endpoints below)
  - For rendering info, use endpoint for getting selected selective disciplines (see endpoints below)

If group and selective disciplines changed simultaneously, make request for editing student, then request for editing student selectives

image

Create student:

- Sidebar that is already implemented
- Button "Cancel" which opens "Search students" page
- Button "Delete" which opens confirmation pop-up. After deleting opens "Search students" page
- Button "Save" which send a request on back-end endpoint which saves data. After saving opens "Search students" page
- Input for first name (min 2, max 40)
- Input for last name (min 2, max 40)
- Input for middle name (min 0, max 40)
- Input for username (min 2, max 40)
- Dropdown for group, multichoice dropdown with an search input
- Dropdown for group role (CAPTAIN, STUDENT, MODERATOR only)
- Button "Edit user" that opens "Edit user" page with user id that has student
- Edit selectives:
  - Each part splitted by semester where user can have selectives:
    - In each part you can add or delete selectives:
      - Bucket icon will delete selectives
      - Button "Add" will show new dropdown, where you can select new discipline. In these dropdown there cannot be disciplines that already picked by the user
      - Also we will send infromation about max amount of selective that user can have in each semester. If limit is reached, do not show "Add" button for exact semester
  - Use endpoint for getting selective disciplines from group (see endpoints below)

image

## Back-end part

- Endpoint for getting students that has an ability to (access is restricted):
  - filter by:
    - group ids array
    - group role array (CAPTAIN, STUDENT, MODERATOR)
    - full name
  - and sort by:
    - group code
    - group role (CAPTAIN, STUDENT, MODERATOR)
    - full name

- Endpoint for editing student (by user id) (access is restricted):
  - first name (string, min 2, max 40)
  - last name (string, min 2, max 40)
  - middle name (string, min 0, max 40)
  - group id:
    - in case of editing group, all selective disciplines must be disconnected from user
    - in case of editing group and user has captain role, you must throw an error
    - in case of editing group and user will have captain role, you need to make already chosen captain to be student
    - in case of editing group, previous group role must be disconnected
  - group role (CAPTAIN, MODERATOR, STUDENT)

- Endpoint for creating student (access is restricted):
  - first name (string, min 2, max 40)
  - last name (string, min 2, max 40)
  - middle name (string, min 0, max 40)
  - username (string, min 2, max 40):
    - if user by username is not found, throw an error
    - user id will be used for creating student entity
  - group id:
    - if user will have captain role, you need to make already chosen captain to be student
  - group role (CAPTAIN, MODERATOR, STUDENT)

- Endpoint for editing student selectives (by user id) (access is restricted):
  - array of connected selective discplines (array of ids)
  - array of disconnected selective disciplines (array of ids)
  - all connected selective disciplines must be check for connecting to student's group

- Endpoint for getting selective disciplines already picked by user (by user id) (access is restricted):
  - array of objects with fields:
    - semester (int)
    - year (int)
    - array of objects with fields:
      - subject name (string)
      - discipline id (uuid)

- Endpoint for getting remaining selective disciplines (by user id) (access is restricted):
  - array of objects with fields:
    - semester (int)
    - year (int)
    - remaining amount (int)
    - array of objects with fields:
      - subject name (string)
      - discipline id (uuid)

- Endpoint for getting all selective disciplines in group (by group id) (access is restricted):
  - array of objects with fields:
    - semester (int)
    - year (int)
    - max amount of selectives (int)
    - array of objects with fields:
      - subject name (string)
      - discipline id (uuid)

- Endpoint for deleting student (by user id)

## UPD
- Endpoint for getting student (by user id)

### PR DESCRIPTION
**Pull request by aim4ik11**
- created admin students flow

closes #87

### CODE REVIEW

**Review by fokaaas — DISMISSED** [review]

File: fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/AdminStudentCreate.styles.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  width: '100%',
+  padding: '24px 0px',
+  margin: '24px 16px',
+};
+
+export const divider: SxProps<Theme> = {
+  height: '1px',
+  width: '100%',
+  marginY: '6px',
+  background: '#747474',
```
Comment (igaryakqwe) [inline fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/AdminStudentCreate.styles.ts:25]: use there backgroundColor: 'backgroundDark.700'

**Review by igaryakqwe — APPROVED** [review]

**Review by fokaaas — APPROVED** [review]

**Review by fokaaas — APPROVED** [review]

File: fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/AdminStudentCreate.tsx [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+  const toast = useToast();
+  const { displayError } = useToastError();
+  const router = useRouter();
+
+  const handleSubmit = useCallback(
+    async (data: StudentCreate) => {
+      try {
+        const response = await StudentAPI.create(data);
+        console.log(response);
```
Comment (igaryakqwe) [inline fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/AdminStudentCreate.tsx:32]: remove console.log

File: fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/AdminStudentCreate.tsx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        if (connectedSelective.length) {
+          console.log(connectedSelective);
```
Comment (igaryakqwe) [inline fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/AdminStudentCreate.tsx:34]: remove console.log

File: fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/components/create-selective/components/dropdowns/Dropdowns.tsx [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  const [pickedSelectives, setPickedSelectives] = useState<
+    Array<{ id: string; name: string }>
+  >([]);
+
+  const [availableAmount, setAvailableAmount] = useState(
+    remainingSelective.amount,
+  );
+  const [isAddingSelective, setIsAddingSelective] = useState(false);
+
+  function deleteSelective(currentDiscipline: DropDownOption) {
```
Comment (igaryakqwe) [inline fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/components/create-selective/components/dropdowns/Dropdowns.tsx:49]: use arrow functions instead function declaration

File: fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/components/create-selective/components/dropdowns/Dropdowns.tsx [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+      return prev.filter(selective => selective.id !== currentDiscipline.id);
+    });
+    setConnectedSelective(prev => {
+      return prev.filter(selective => selective !== currentDiscipline.id);
+    });
+    setSelectiveOptions(options => [...options, currentDiscipline]);
+    setAvailableAmount(prev => prev + 1);
+  }
+
+  function editSelective(newValue: string, oldValue: DropDownOption) {
```
Comment (igaryakqwe) [inline fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/components/create-selective/components/dropdowns/Dropdowns.tsx:60]: use arrow functions instead function declaration

File: fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/components/create-selective/components/dropdowns/Dropdowns.tsx [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    ]);
+    // change available options accordingly
+    const newRemainingSelectives = selectiveOptions.filter(
+      selective => selective.id !== newValue,
+    );
+    newRemainingSelectives.push(oldValue);
+    setSelectiveOptions(newRemainingSelectives);
+  }
+
+  function handleSelectiveAdd(newValue: string) {
```
Comment (igaryakqwe) [inline fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/components/create-selective/components/dropdowns/Dropdowns.tsx:74]: use arrow functions instead function declaration

File: fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/components/create-selective/components/dropdowns/Dropdowns.tsx [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+          </Box>
+        ) : (
+          <Dropdown
+            size={FieldSize.SMALL}
+            options={selectiveOptions}
+            showRemark={false}
+            onChange={(value: string) => {
+              handleSelectiveAdd(value);
+            }}
+            value={''}
```
Comment (igaryakqwe) [inline fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/components/create-selective/components/dropdowns/Dropdowns.tsx:125]: wtf

File: fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/components/create-selective/components/dropdowns/Dropdowns.tsx [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+          />
+        ))}
+      {!!availableAmount &&
+        (!isAddingSelective ? (
+          <Box sx={styles.addButton}>
+            <Button
+              color={ButtonColor.PRIMARY}
+              variant={ButtonVariant.OUTLINE}
+              size={ButtonSize.SMALL}
+              text={'Додати'}
```
Comment (igaryakqwe) [inline fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/components/create-selective/components/dropdowns/Dropdowns.tsx:112]: remove curly braces

File: fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/components/create-selective/components/dropdowns/components/SelectiveDropdown.tsx [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        options={[...options, currentDiscipline]}
+        showRemark={false}
+        onChange={(value: string) => {
+          editSelective(value, currentDiscipline);
+          const newOption = options.find(option => option.id === value);
+          setCurrentDiscipline(newOption!);
+        }}
+        value={currentDiscipline.id}
+        label=""
+        placeholder=""
```
Comment (igaryakqwe) [inline fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/components/create-selective/components/dropdowns/components/SelectiveDropdown.tsx:47]: do you need to write it there?

File: fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/components/create-student-inputs/CreateStudentInputs.tsx [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        label="Username"
+        placeholder="Username"
+      />
+      <FormikDropdown
+        name="groupId"
+        disableClearable
+        placeholder="Група"
+        size={FieldSize.LARGE}
+        options={groupsOptions}
+        label=""
```
Comment (igaryakqwe) [inline fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/components/create-student-inputs/CreateStudentInputs.tsx:55]: do you need to write it there?

File: fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/components/create-student-inputs/CreateStudentInputs.tsx [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+      />
+      <FormikDropdown
+        name="roleName"
+        disableClearable
+        placeholder="Роль"
+        size={FieldSize.LARGE}
+        options={StudentRoleOptions}
+        label=""
```
Comment (igaryakqwe) [inline fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/components/create-student-inputs/CreateStudentInputs.tsx:63]: do you need to write it there?

File: fictadvisor-web/src/components/pages/admin/admin-student/edit-student-page/components/edit-selective/components/dropdowns/Dropdowns.tsx [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+  );
+  const [availableAmount, setAvailableAmount] = useState(
+    remainingSelective.availableSelectiveAmount,
+  );
+  const [isAddingSelective, setIsAddingSelective] = useState(false);
+  const [pickedSelectives, setPickedSelectives] = useState<
+    Array<{ id: string; name: string }>
+  >(currentSelective?.selective || []);
+
+  function editSelective(newValue: string, oldValue: DropDownOption) {
```
Comment (igaryakqwe) [inline fictadvisor-web/src/components/pages/admin/admin-student/edit-student-page/components/edit-selective/components/dropdowns/Dropdowns.tsx:38]: use arrow functions

File: fictadvisor-web/src/components/pages/admin/admin-student/edit-student-page/components/edit-selective/components/dropdowns/Dropdowns.tsx [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+    });
+    // change available options accordingly
+    const newSelectiveOptions = selectiveOptions.filter(
+      selective => selective.id !== newValue,
+    );
+    newSelectiveOptions.push(oldValue);
+    setSelectiveOptions(newSelectiveOptions);
+  }
+
+  function handleSelectiveAdd(newValue: string) {
```
Comment (igaryakqwe) [inline fictadvisor-web/src/components/pages/admin/admin-student/edit-student-page/components/edit-selective/components/dropdowns/Dropdowns.tsx:62]: use arrow functions

File: fictadvisor-web/src/components/pages/admin/admin-student/edit-student-page/components/edit-selective/components/dropdowns/Dropdowns.tsx [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+    // remove it from available options
+    setSelectiveOptions(options =>
+      options.filter(selective => selective.id !== newValue),
+    );
+    // change variables
+    setAvailableAmount(prev => prev - 1);
+    setIsAddingSelective(false);
+  }
+
+  function handleSelectiveDelete(value: DropDownOption) {
```
Comment (igaryakqwe) [inline fictadvisor-web/src/components/pages/admin/admin-student/edit-student-page/components/edit-selective/components/dropdowns/Dropdowns.tsx:90]: use arrow functions

File: fictadvisor-web/src/components/pages/admin/admin-student/edit-student-page/components/edit-selective/components/dropdowns/Dropdowns.tsx [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+          />
+        ))}
+      {!!availableAmount &&
+        (!isAddingSelective ? (
+          <Box sx={styles.addButton}>
+            <Button
+              color={ButtonColor.PRIMARY}
+              variant={ButtonVariant.OUTLINE}
+              size={ButtonSize.SMALL}
+              text={'Додати'}
```
Comment (igaryakqwe) [inline fictadvisor-web/src/components/pages/admin/admin-student/edit-student-page/components/edit-selective/components/dropdowns/Dropdowns.tsx:132]: remove curly braces

File: fictadvisor-web/src/components/pages/admin/admin-student/edit-student-page/components/edit-selective/components/dropdowns/Dropdowns.tsx [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+              onClick={() => setIsAddingSelective(true)}
+            />
+          </Box>
+        ) : (
+          <Dropdown
+            size={FieldSize.SMALL}
+            options={selectiveOptions}
+            showRemark={false}
+            onChange={(value: string) => handleSelectiveAdd(value)}
+            value={''}
```
Comment (igaryakqwe) [inline fictadvisor-web/src/components/pages/admin/admin-student/edit-student-page/components/edit-selective/components/dropdowns/Dropdowns.tsx:143]: remove curly braces

File: fictadvisor-web/src/components/pages/admin/admin-student/edit-student-page/components/edit-selective/components/dropdowns/components/SelectiveDropdown.tsx [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+        options={[...options, currentDiscipline]}
+        showRemark={false}
+        onChange={(value: string) => {
+          editSelective(value, currentDiscipline);
+          const newOption = options.find(option => option.id === value);
+          setCurrentDiscipline(newOption!);
+        }}
+        value={currentDiscipline.id}
+        label=""
+        placeholder=""
```
Comment (igaryakqwe) [inline fictadvisor-web/src/components/pages/admin/admin-student/edit-student-page/components/edit-selective/components/dropdowns/components/SelectiveDropdown.tsx:47]: ?

**Review by igaryakqwe — COMMENTED** [review]
remove all comments

File: fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/components/create-student-inputs/CreateStudentInputs.tsx (hunk 11 above)
Comment (sifes) [reply fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/components/create-student-inputs/CreateStudentInputs.tsx:63]: yes, i do

File: fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/components/create-student-inputs/CreateStudentInputs.tsx (hunk 10 above)
Comment (sifes) [reply fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/components/create-student-inputs/CreateStudentInputs.tsx:55]: yes, i do

File: fictadvisor-web/src/components/pages/admin/admin-student/edit-student-page/components/edit-selective/components/dropdowns/components/SelectiveDropdown.tsx (hunk 17 above)
Comment (sifes) [reply fictadvisor-web/src/components/pages/admin/admin-student/edit-student-page/components/edit-selective/components/dropdowns/components/SelectiveDropdown.tsx:47]: yes.

File: fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/components/create-selective/components/dropdowns/components/SelectiveDropdown.tsx (hunk 9 above)
Comment (sifes) [reply fictadvisor-web/src/components/pages/admin/admin-student/create-student-page/components/create-selective/components/dropdowns/components/SelectiveDropdown.tsx:47]: yes, i do

**Pull request closed without merging**